modules/data.atmosphere/R/tdm_generate_subdaily_models.R

```diff
@@ -0,0 +1,226 @@
+##' Create statistical models to predict subdaily meteorology from daily means
+##' @family tdm - Temporally Downscale Meteorology
+##' @title gen_subdaily_models
+##' @param outfolder - directory where models will be stored *** storage required varies by size of training dataset, but prepare for >100 GB
+##' @param dat.train_file - train_data file
+##' @param in.prefix 
+##' @param tdf_file - tdm_temporal_downscale_functions.R filepath that can be sourced i.e. '~/scripts/temporal_downscale_functions.R'
+##' @param n.beta - number of betas to save from linear regression model
+##' @param resids - logical stating whether to pass on residual data or not
+##' @param parallel - logical stating whether to run temporal_downscale_functions.R in parallel 
+##' @param n.cores - deals with parallelization
+##' @param day.window - integer specifying number of days around a particular day you want to use data from for that 
+##'                     specific hours coefficients
+##' @param overwrite
+##' @param verbose
+
+##' @author Christy Rollinson, James Simkins
+
+#----------------------------------------------------------------------
+# Begin Script
+#----------------------------------------------------------------------
+
+gen_subdaily_models <- function(outfolder, dat.train_file, in.prefix, tdf_file, 
+    n.beta, day.window, resids = F, parallel = F, n.cores = NULL, overwrite = TRUE, 
+    verbose = FALSE) {
+
+    pb.index <- 1
+    pb <- txtProgressBar(min = 1, max = 8, style = 3)
+    
+    # ----- 1.0 Read data & Make time stamps ---------- Load the data
+    
+    vars.info <- data.frame(CF.name = c("date", "year", "doy", "hour", 
+        "air_temperature", "precipitation_flux", "air_temperature_max", 
+        "air_temperature_min", "surface_downwelling_shortwave_flux_in_air", 
+        "surface_downwelling_longwave_flux_in_air", "air_pressure", "specific_humidity", 
+        "eastward_wind", "northward_wind", "wind_speed"))
+    dat.train <- list()
+    tem <- ncdf4::nc_open(dat.train_file)
+    dim <- tem$dim
+    for (j in seq_along(vars.info$CF.name)) {
+        if (exists(as.character(vars.info$CF.name[j]), tem$var) == FALSE) {
+            dat.train[[j]] <- NA
+        } else {
+            dat.train[[j]] <- ncdf4::ncvar_get(tem, as.character(vars.info$CF.name[j]))
+        }
+    }
+    names(dat.train) <- vars.info$CF.name
+    dat.train <- data.frame(dat.train)
+    
+    dat.train$date <- strptime(paste(dat.train$year, dat.train$doy + 1, 
+        dat.train$hour, sep = "-"), "%Y-%j-%H", tz = "GMT")
+    dat.train$time.hr <- as.numeric(difftime(dat.train$date, paste0((min(dat.train$year) - 
+        1), "-12-31 23:00:00"), tz = "GMT", units = "hour"))
+    dat.train$time.day <- as.numeric(difftime(dat.train$date, paste0((min(dat.train$year) - 
```

Comment:
These are used in the linear regression model. We need to keep our time of days consistent. I'll document this better